R/03a-create-catchments.R

```diff
@@ -0,0 +1,164 @@
+
+# Locations of all Service BC locations 
+sbc_locs <- read_csv(glue("{SRC_DATA_FOLDER}/full-service-bc-locs.csv")) |>
+  st_as_sf(
+    coords = c('coord_x', 'coord_y'),
+    crs = 3005
+  ) 
+
+# Drive time data - required for 'SBC catchments'
+drivetime_data <-
+  read_csv(
+    glue("{SRC_DATA_FOLDER}/full-processed-drivetime-data.csv"),
+    col_types = cols(.default = "c")
+  ) %>%
+  clean_names() %>%
+  mutate(across(c(drv_time_sec, drv_dist), as.numeric))
+
+# DB shapefiles
+db_shapefile <-
+  st_read(glue("{SHAPEFILE_OUT}/full-db_with_location.gpkg")) %>%
+  mutate(across(c(landarea), as.numeric))
+
+# list of addresses with no valid routes
+db_no_route <- read_csv(
+    glue("{DT_DATA_FOLDER}/final_result_only_errors.csv")
+    ) %>% 
+    select(dbid = DISSEMINATION_BLOCK_ID) %>% 
+    mutate(
+        dbid = as.character(dbid),
+        no_route = TRUE
+        )  %>% 
+    distinct()
+
+# ----------------------------------------------------------------------------
+# Create catchment areas using traditional drive time assignment
+# ----------------------------------------------------------------------------
+
+# First assign every dissemination block a 'nearest facility' based on drive time
+drive_time_assignments <- assign_nearest_facility(drivetime_data)
+
+# Calculate how many DBs are unassigned
+unassigned_count_before <- db_shapefile |>
+  left_join(drive_time_assignments, by = "dbid") |>
+  filter(is.na(assigned)) |>
+  nrow()
+
+message("Number of unassigned DBs before spatial assignment: ", unassigned_count_before)
+
+# ----------------------------------------------------------------------------
+# Apply spatial assignment method to fill in unassigned areas
+# ----------------------------------------------------------------------------
+
+# Assign all unassigned DBs directly to the nearest facility
+complete_assignments <- assign_unassigned_dbs(
+  db_shapefile, 
+  drive_time_assignments,
+  facility_locations = sbc_locs,
+  batch_size = 20000,
+  verbose = TRUE
+)
+
+# Calculate how many DBs are unassigned after spatial assignment
+unassigned_count_after <- db_shapefile |>
+  left_join(complete_assignments, by = "dbid") |>
+  filter(is.na(assigned)) |>
+  nrow()
+
+message("Number of unassigned DBs after spatial assignment: ", unassigned_count_after)
+
+# Calculate how many DBs were assigned using the spatial method
+spatial_assignments_count <- complete_assignments |>
+  filter(assignment_method == "nearest_facility") |>
+  nrow()
+
+message("Number of DBs assigned directly to nearest facility: ", spatial_assignments_count)
+
+# Save the complete assignments for use in other scripts
+write_csv(
+  complete_assignments,
+  glue("{SRC_DATA_FOLDER}/complete_db_assignments.csv")
+)
+
+# QA the unassigned DBs 
+# save the list of unassigned dbs, together with their 2021 census pops
+db_check <- complete_assignments %>% 
+  left_join(pop_db, by='dbid') %>% 
```

Comment:
pop_db hasn't been defined in this script yet

Reply:
updated!